### PR TITLE
Fold mul (lshr exact (X, N)), 2^N + 1 to add (X , lshr exact (X, N))

### PR DESCRIPTION
**Pull request by AZero13**
Alive2 Proofs:
https://alive2.llvm.org/ce/z/aJnxyp
https://alive2.llvm.org/ce/z/dyeGEv

### CODE REVIEW

**Comment by AtariDreams** [comment]
@dtcxzyw Thoughts on this?

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 1]
```diff
@@ -255,6 +255,43 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
     }
   }
 
+  {
+    // mul (lshr exact (X, 2^N + 1)), N -> add (X , lshr (X, N))
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:259]:
```suggestion
    // mul (lshr exact X, N), (2^N + 1) -> add (X , lshr (X, N))
```

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 2]
```diff
@@ -255,6 +255,43 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
...
+            AddOp = NewOp;
+
+          Value *BinOp;
+          if (OpBO->getOpcode() == Instruction::LShr ||
+              (OpBO->getOpcode() == Instruction::AShr && HasNUW)) {
+            BinOp = Builder.CreateLShr(
+                AddOp, ConstantInt::get(Ty, ShiftC->getZExtValue()), "", true);
+          } else {
+            BinOp = Builder.CreateAShr(
+                AddOp, ConstantInt::get(Ty, ShiftC->getZExtValue()), "", true);
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:284]: comment the `true`.

File: llvm/test/Transforms/InstCombine/ashr-lshr.ll [hunk 3]
```diff
@@ -862,4 +862,130 @@ define i32 @ashr_mul_times_5_div_4_exact_2(i32 %x) {
...
+; CHECK-LABEL: @ashr_shift_mul_nsw(
+; CHECK-NEXT:    [[TMP1:%.*]] = freeze i32 [[X:%.*]]
+; CHECK-NEXT:    [[TMP2:%.*]] = ashr exact i32 [[TMP1]], 3
+; CHECK-NEXT:    [[RES:%.*]] = add nsw i32 [[TMP1]], [[TMP2]]
+; CHECK-NEXT:    ret i32 [[RES]]
+;
+  %a = ashr exact i32 %x, 3
+  %res = mul nsw i32 %a, 9
+  ret i32 %res
+}
```
Comment (goldsteinn) [inline llvm/test/Transforms/InstCombine/ashr-lshr.ll:898]: Is this case correct?
https://alive2.llvm.org/ce/z/Vag8Wh

Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/ashr-lshr.ll:898]: It's not. I'm gonna fix it today.

**Comment by AtariDreams** [comment]
Okay so if the IR just repeats I guess I can remove one use but what if there's frozen involved @goldsteinn 

Then it should be one use no?

**Comment by AtariDreams** [comment]
> > Okay so if the IR just repeats I guess I can remove one use but what if there's frozen involved @goldsteinn
> 
> > 
> 
> > Then it should be one use no?
> 
> 
> 
> Im not sure what you are asking.

See:
image

**Comment by goldsteinn** [comment]
> > > Okay so if the IR just repeats I guess I can remove one use but what if there's frozen involved @goldsteinn
> > 
> > 
> > > 
> > 
> > 
> > > Then it should be one use no?
> > 
> > 
> > Im not sure what you are asking.
> 
> See:


As a general rule, don't create more instructions. So in that case, req oneuse.

**Comment by AtariDreams** [comment]
> > > > Okay so if the IR just repeats I guess I can remove one use but what if there's frozen involved @goldsteinn
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > Then it should be one use no?
> 
> > > 
> 
> > > 
> 
> > > Im not sure what you are asking.
> 
> > 
> 
> > See:
> 
> 
> 
> 
> 
> As a general rule, don't create more instructions. So in that case, req oneuse.

And what about the freeze? Does that count?

**Comment by AtariDreams** [comment]
> > > > Okay so if the IR just repeats I guess I can remove one use but what if there's frozen involved @goldsteinn
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > Then it should be one use no?
> 
> > > 
> 
> > > 
> 
> > > Im not sure what you are asking.
> 
> > 
> 
> > See:
> 
> 
> 
> 
> 
> As a general rule, don't create more instructions. So in that case, req oneuse.

I'll just reuse Op0 directly I think hopefully it works

File: llvm/test/Transforms/InstCombine/ashr-lshr.ll (hunk 3 above)
Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/ashr-lshr.ll:897]: I don't think you have solved the problem that you can't propagate `nsw` to the add if the ShiftAmt may be `BitWidth - 1` or `MulAp` may be INT_MIN. This case might get simplified beforehand, but it still shouldn't break w/ your code.

Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/ashr-lshr.ll:896]: Fixed!

**Comment by goldsteinn** [comment]
Okay, this LGTM. Please wait on an additional approval to push.

**Comment by AtariDreams** [comment]
@nikic Thoughts?

**Comment by AtariDreams** [comment]
@nikic Can we please merge this?

**Comment by AtariDreams** [comment]
@dtcxzyw I do not have merge permissions.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 4]
```diff
@@ -255,6 +255,38 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
...
   }
 
+  {
+    // mul (lshr exact X, N), (2^N + 1) -> add (X , lshr (X, N))
+    Value *NewOp;
+    const APInt *ShiftC;
+    const APInt *MulAP;
+    if ((HasNSW || HasNUW) &&
+        match(&I, m_Mul(m_CombineOr(m_LShr(m_Value(NewOp), m_APInt(ShiftC)),
+                                    m_AShr(m_Value(NewOp), m_APInt(ShiftC))),
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:265]:
```suggestion
        match(&I, m_Mul(m_Shr(m_Value(NewOp), m_APInt(ShiftC)),
```

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 5]
```diff
@@ -255,6 +255,38 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
...
+                        m_APInt(MulAP)))) {
+      if (BitWidth > 2 && (*MulAP - 1).isPowerOf2() &&
+          MulAP->logBase2() == ShiftC->getZExtValue()) {
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:268]:
```suggestion
          ShiftC == MulAP->logBase2()) {
```

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 6]
```diff
@@ -255,6 +255,38 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
...
+        BinaryOperator *OpBO = cast<BinaryOperator>(Op0);
+        if (OpBO->isExact()) {
+          Value *BinOp = Op0;
+          if (HasNUW && OpBO->getOpcode() == Instruction::AShr &&
+              OpBO->hasOneUse())
+            BinOp = Builder.CreateLShr(
+                NewOp, ConstantInt::get(Ty, ShiftC->getZExtValue()), "",
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:275]:
```suggestion
                NewOp, ConstantInt::get(Ty, ShiftC), "",
```

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 7]
```diff
@@ -255,6 +255,38 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
...
+    const APInt *ShiftC;
+    const APInt *MulAP;
+    if ((HasNSW || HasNUW) &&
+        match(&I, m_Mul(m_CombineOr(m_LShr(m_Value(NewOp), m_APInt(ShiftC)),
+                                    m_AShr(m_Value(NewOp), m_APInt(ShiftC))),
+                        m_APInt(MulAP)))) {
+      if (BitWidth > 2 && (*MulAP - 1).isPowerOf2() &&
+          MulAP->logBase2() == ShiftC->getZExtValue()) {
+        BinaryOperator *OpBO = cast<BinaryOperator>(Op0);
+        if (OpBO->isExact()) {
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:270]: Use `m_Exact` in the pattern match instead.

**Review by nikic — COMMENTED** [review]
The code doesn't match the proofs (the proofs have freeze instructions, the code does not).

**Comment by AtariDreams** [comment]
The freeze instructions added more instructions so I removed handling for that and now nsw or nuw is required. Freeze is for when none of those are present.

**Comment by AtariDreams** [comment]
> The code doesn't match the proofs (the proofs have freeze instructions, the code does not).

Is it okay to use freeze even if that counts as an extra instruction in the IR? Because that's why did a last second replacement with a check for NUW/NSW for the optimization to apply at all.

**Comment by AtariDreams** [comment]
> The code doesn't match the proofs (the proofs have freeze instructions, the code does not).

Fixed!

**Comment by AtariDreams** [comment]
@dtcxzyw @nikic What do you think about this?

**Comment by RSilicon** [comment]
Okay we are back!

**Comment by RSilicon** [comment]
> Okay, this LGTM. Please wait on an additional approval to push.

Thoughts?

**Comment by RSilicon** [comment]
@dtcxzyw @goldsteinn Is this good now?

**Comment by DTeachs** [comment]
@nikic Did you see this yet?

**Comment by RSilicon** [comment]
W

**Comment by RSilicon** [comment]
@dtcxzyw as I was saying, would you please take a look and approve this please.

**Comment by RSilicon** [comment]
Closing for now due to lack of interest from maintainers.

**Comment by AreaZR** [comment]
@nikic can we please merge this and be done with it?

**Comment by AreaZR** [comment]
@goldsteinn Thoughts?

**Comment by AreaZR** [comment]
@goldsteinn @dtcxzyw Thoughts?

**Comment by AreaZR** [comment]
@dtcxzyw, could you please review and approve this? Thank you!

**Comment by AreaZR** [comment]
@nikic Could you please review this when you get the chance? Thank you.

**Pull request closed without merging**